exporter/otlphttpexporter/otlp.go

```diff
@@ -199,29 +203,102 @@ func isRetryableStatusCode(code int) bool {
 	}
 }
 
+func readResponseBody(body io.ReadCloser) ([]byte, error) {
+	// Read the maximum number of bytes allowed in a request. This avoids
+	// issues with missing or invalid Content-Length headers.
+	protoBytes := make([]byte, maxHTTPResponseReadBytes)
+	n, err := io.ReadFull(body, protoBytes)
+
+	// No bytes read and an EOF error indicates there is no body to read.
+	if n == 0 && (err == nil || errors.Is(err, io.EOF)) {
+		return nil, nil
+	}
+
+	// io.ReadFull will return io.ErrorUnexpectedEOF in most cases since there
+	// will usually be a mismatch between the length of the byte slice and the
+	// size of the body, so we ignore that error.
+	if err != nil && !errors.Is(err, io.ErrUnexpectedEOF) {
+		return nil, err
+	}
+
+	// The pdata unmarshaling methods check for the length of the slice
+	// when unmarshaling it, so we have to trim down the length to the
+	// actual size of the data.
+	return protoBytes[:n], nil
+}
+
 // Read the response and decode the status.Status from the body.
 // Returns nil if the response is empty or cannot be decoded.
-func readResponse(resp *http.Response) *status.Status {
+func readResponseStatus(resp *http.Response) *status.Status {
```

Comment:
Are changes in this function related to handling partial request?

Reply:
Not directly, this was a change requested in earlier feedback around some code duplication that was introduced: https://github.com/open-telemetry/opentelemetry-collector/pull/6970#discussion_r1143536486. We can make this change in a follow-up PR if you'd prefer, I'm fine either way.